### PR TITLE
Exclude default values from Route Values

### PR DESCRIPTION
**Pull request by russcam**
PR for https://github.com/T4MVC/T4MVC/issues/44

Introduction of a `SanitizeDefaultValue(string defaultValue, string type)` method that takes the default value string and type of an action method parameter and attempts a regex match for
1. `default(T)` for some type T
2. parameterless constructor for a type (valid optional parameter for a Value type)

if a match is found, the match is replaced with a corresponding string that uses the fully qualified type name, and this value is compared against the value for an action parameter to determine if the value should be added to the `RouteValueDictionary` and used in rendering a url for the controller action.

For example,

`default(CancellationToken)` is replaced with `default(System.Threading.CancellationToken)`
`default(    CancellationToken   )` is replaced with `default(System.Threading.CancellationToken)`
`new CancellationToken()` is replaced with `new System.Threading.CancellationToken()`

with a controller action such as

```
public partial class HomeController : Controller
{
    public virtual async Task<ActionResult> Index(CancellationToken cancellationToken = default(CancellationToken))
    {
        // perform some work that uses the cancellation token
        await Task.Delay(1000, cancellationToken);

        return View();
    }
}
```

when a T4MVC overload such as `Url.Action(MVC.Home.Index())` is used to generate a url for the controller action, a query string parameter is not appended to the generated url for the default value of the cancellation token passed as an optional parameter.

### CODE REVIEW

File: T4MVCHostMvcApp/T4MVC Files/HomeController.generated.cs [hunk 1]
```diff
@@ -521,12 +564,65 @@ public override System.Web.Mvc.ActionResult ActionWithSomeOptionalParams(string
...
+        }
+
+        [NonAction]
+        partial void ActionWithOptionalParamDefaultingToConstructedDefaultValueOverride(T4MVC_System_Web_Mvc_ActionResult callInfo, System.Threading.CancellationToken cancellationToken);
+
+        [NonAction]
+        public override System.Web.Mvc.ActionResult ActionWithOptionalParamDefaultingToConstructedDefaultValue(System.Threading.CancellationToken cancellationToken)
+        {
+            var callInfo = new T4MVC_System_Web_Mvc_ActionResult(Area, Name, ActionNames.ActionWithOptionalParamDefaultingToConstructedDefaultValue);
+            if (cancellationToken != new System.Threading.CancellationToken())
```
Comment (davidebbo) [inline T4MVCHostMvcApp/T4MVC Files/HomeController.generated.cs]: I don't get the `new` case. How could that ever be equal? `new Foo()` is never equal to `new Foo()`.

**Comment by davidebbo** [comment]
One comment about the `new case` inline.

Also, note that I rebased all the work on top for the latest master and resolved minor conflicts. You also had some tabs that sneaked in the .tt that I fixed :)

**Important**: to make additional changes, please make sure to start from my rebased `NoDefaultParams` branch and not the one you currently use. Thanks!

**Comment by russcam** [comment]
Thanks for fixing the tabs - will update my IDE settings :)

According to the C# (5) Specification, _Section 10.6.1 Method Parameters_,

> The expression in a default-argument must be one of the following:
> - a constant-expression 
> - an expression of the form `new S()` where `S` is a value type
> - an expression of the form `default(S)` where `S` is a value type
> 
> The expression must be implicitly convertible by an identity or nullable conversion to the type of the parameter.

So `new S()` will only ever apply to value types. Additionally, [`ValueType.Equals(Object)` overrides `Object.Equals(Object)`](https://msdn.microsoft.com/en-us/library/2dts52z7.aspx):

> The `ValueType.Equals(Object)` method overrides `Object.Equals(Object)` and provides the default implementation of value equality for all value types in the .NET Framework.
> **If none of the fields of the current instance and obj are reference types, the Equals method performs a byte-by-byte comparison of the two objects in memory. Otherwise, it uses reflection to compare the corresponding fields of obj and this instance.**

So, for a value type that does not have any fields that are reference types,

```
new Foo().Equals(new Foo()) == true
```

Hopefully a developer would override `.Equals(Object)` for a value type with a reference type field as MSDN recommends :)

With C# 6, the parameterless constructor can be called for a value type even when it is not explicitly defined. For example

```
public static void Main()
{
    CompareDefaultFoosForEquality();
    CompareNewFoosForEquality();
    CompareDefaultAndNewFoosForEquality();
}

public static void CompareDefaultFoosForEquality(Foo foo = default(Foo))
{
    Console.WriteLine("default(Foo).Equals(default(Foo): {0}", foo.Equals(default(Foo)));
}

public static void CompareNewFoosForEquality(Foo foo = new Foo())
{
    Console.WriteLine("new Foo().Equals(new Foo()): {0}", foo.Equals(new Foo()));
}

public static void CompareDefaultAndNewFoosForEquality(Foo foo = default(Foo))
{
    Console.WriteLine("default(Foo).Equals(new Foo()): {0}", foo.Equals(new Foo()));
}

public struct Foo
{
    public Foo(int x, int y)
    {
        X = x;
        Y = y;
    }

    public int X { get; private set; }
    public int Y { get; private set; }
}
```

yields 

> default(Foo).Equals(default(Foo): **True**
> new Foo().Equals(new Foo()): **True**
> default(Foo).Equals(new Foo()): **True**

Based on this, I think the Equality check for `new` with value types and using `.Equals()` is good, with the exception of `default(T)` and `new T()` [where the parameterless constructor initializes fields to non-zero values](http://www.volatileread.com/Wiki/Index?id=1091), but I don't think this is a concern here as the code sanitizes what the user has supplied as the default value for the action parameter and hence is a sanitized reflection of their input.

This has however raised a different problem. The code currently uses `==` to check for equality, but this will only work for a value type if the `==` operator is overloaded; if the operator has not been overloaded, the following compiler error is exhibited

> CS0019 Operator '==' cannot be applied to operands of type 'T' and 'T'

In summary, I think the `new` check is OK and the equality check should use `.Equals(Object)` instead of `==`. What are your thoughts? Are there any ramifications to changing `==` to `.Equals(Object)` that you can think of?

**Comment by davidebbo** [comment]
Thanks for the detailed explanation. This makes sense! I changed it to `.Equals()` (c6c74cbcbfe83b6c0a883bbeb8d49ca4b94c6edc) and published new version 3.16.1. Thanks for your contrib!

**Comment by russcam** [comment]
Awesome, thank you @davidebbo!
